---
title: "Brobustedger"
author: "Bandhan Sarker, Md. Matiur Rahaman, Muhammad Habibulla Alamin, Md. Ariful Islam, and Md. Nurul Haque Mollah"
date: "`r Sys.Date()`"
abstract: >
  **Brobustedger** is an R package for identifying differentially expressed genes (DEGs)
  from RNA-seq count data in the presence of gene-specific outliers and missing values.
  The package integrates an iterative leave-one-out (iLOO) strategy for detecting
  gene-level outliers and a random forestâ€“based approach for imputing missing or null
  values using the **missRanger** package. Differential expression analysis is then
  performed using robust methods implemented in **edgeR (Robust)**. Brobustedger provides a
  user-friendly interface through the **DEGList** function, along with standalone
  utilities for outlier detection (**iLOO**) and missing value imputation (**NI**).
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Brobustedger}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
Alternatively, you can install the development version directly from GitHub using `devtools`:


## Installation
```{r github, eval=FALSE}
devtools::install_github("BandhanSarker/Brobustedger")
```

And load `Brobustedger`:
```{r}
library(Brobustedger)
```

## Data Pre-process
## Step-1:
iLOO is used to identify gene specific outliers from RNA-seq count data.
## Usage
```{r iloo-usage, eval=FALSE}
head(iLOO(data))
```

## Arguments
* **data** RNA-seq count data matrix, whose row contain genes and column contain samples
## Value
Output of the function identify gene specific outliers. Non-outlier values are displayed as "NA"
## Examples
```{r}
head(Brobustedger::iLOO(sampleDataOut)) # Input RNA-seq count data
```
## Step-2:
NI is used for the imputation of null/missing values in RNA-seq count data.
## Usage
```{r , eval=FALSE}
NI(data, mi = 5, NT = 100)
```
## Arguments
* **data**	RNA-seq count data matrix contains missing/Null value, whose row contain genes and column contain samples
* **mi**	maximum number of iterations to impute nullify values
* **NT**    number of trees in each set to impute nullify values

## Value
Clean dataset after the imputation of missing/nullify value

## Examples
```{r}
head(Brobustedger::NI(sampleDataMiss))  #Input RNA-seq count data with missing or null value
```
## Step-3:
DEGList is used to identify differentially expressed genes from RNA-seq count data.
## Usage
```{r , eval=FALSE}
#Identify differentially expressed gene list
DEGList (data, n1, n2, p.threshold = 0.05)
```
## Arguments
* **data**	  RNA-seq count data matrix, whose row contains genes and column contains  samples                 
* **n1**	  control group
* **n2**	  case group
* **p.threshold**	adjusted p-value by default=0.05
* **padjust**	select the p-value adjustment method by default="BH". choose the possible method "none", "BH", "fdr", "BY" and "holm". See help section of BrobustedgeR-package for more details.
* **mi** maximum number of iterations to impute nullify values by default= 5
* **NT** number of trees in each set to impute nullify values by default=10
  
## Value 
1. Gene.list: Differentially expressed genes list and 2. result: others output
## Examples
```{r}
 data (sampleDataOut)
 result.DE<- DEGList(sampleDataOut,3,3,0.05)
 result.matrix<-result.DE$result
 head(result.matrix) #
 ##### DE geneList
 head(result.DE$Gene.list)
```

## Example Dataset
```{r}
#Sample RNA-seq count data with outliers value
data(sampleDataOut)
```
```{r}
#Sample RNA-seq count data with missing value
data(sampleDataMiss)
```
 
## Author(s): 
Bandhan Sarker, Md. Matiur Rahaman, Muhammad Habibulla Alamin, Md. Ariful Islam and Md.  Nurul Haque Mollah

## References
Sarker B, Rahaman MM, Alamin MH et al. (2024) Boosting edgeR (Robust) by dealing with missing observations and gene-specific outliers in RNA-Seq profiles and its application to explore biomarker genes for diagnosis and therapies of ovarian cancer. Genomics. https://doi.org/10.1016/j.ygeno.2024.110834

Robinson MD, McCarthy DJ and Smyth GK (2010). edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics 26, 139-140

Stekhoven, D.J. and Buehlmann, P. (2012), 'MissForest - nonparametric missing value imputation for mixed-type data', Bioinformatics, 28(1) 2012, 112-118, doi: 10.1093/bioinformatics/btr597

George NI, Bowyer JF, Crabtree NM, Chang C-W (2015) An Iterative Leave-One-Out Approach to Outlier Detection in RNA-Seq Data. PLoS ONE 10(6): e0125224. https://doi.org/10.1371/journal.pone.0125224
